Random Draws For SEIR Parameters Per Slot

TimothyWillard opened the pull request:
### Describe your changes.

This pull request addresses the bug described in GH-406 where new seir parameters were not drawn across slots.

### Does this pull request make any user interface changes? If so please describe.

The user interface changes are:

* Added a new `Parameters.reinitialize_distributions` method that recreates the randomly drawn seir parameters so that they do not hold onto the prior random state, and
* `SEIR.run_parallel_SEIR` will now draw random seir parameters across jobs.

These changes have not been documented yet, @alsnhll where do you think is the best place to describe this behavior (I'll make these changes before marking this PR as ready for review)?

### What does your pull request address? Tag relevant issues.

This pull request addresses GH-406.

### Tag relevant team members.

@alsnhll

alsnhll commented:
I did some testing of how random draws of parameters work for each parameter type, using the config config_sample_2pop_modifiers_test_random.yml which is now in the tutorials folder

In the original code in the main branch, the behavior is:
- `spar` file, `seir:parameters:Ro` -  same value by slot - Not working as intended
- `hpar` file, `outcomes:incidDeath:delay` - varying by slot and by location. Only taking on integar values. 
- `snpi` file, `seir_modifier:modifiers:Ro_relax` - varying by slot and by location. 
- `hnpi` file, `outcome_modifiers:modifiers:test_limits` - varying by slot and by location. 

Thus variation in `spar` was the only one _not_ working as intended, while variation in `hpar, snpi`, and `hpi` seem to be working as intended. 

In this branch, the behavior is:
- `spar` file, `seir:parameters:Ro` -  varying by slot
- `hpar` file, `outcomes:incidDeath:delay` - varying by slot and by location. Only taking on integar values. 
- `snpi` file, `seir_modifier:modifiers:Ro_relax` - varying by slot and by location. 
- `hnpi` file, `outcome_modifiers:modifiers:test_limits` - varying by slot and by location. 

So now variation in `spar `seems to be working as intended. 

However note that users are likely surprised by the variation by location in `hpar`, `snpi`, and `hnpi` here, as I don't think this behavior is documented anywhere. It's likely not clear to most users that these parameters have different values by location by default when they randomly vary, even though when their value is fixed they have the same value by location. Is this what we want the default behavior to be? I know during inference values are fit by location by default but does that make sense as the default for forward simulations? It's also confusing that spar parameters don't vary by location while these other ones do (I know that's a long standing flepimop issue but it re-confuses me every time I encounter it). 

If I instead run a config where for both seir and outcome modifiers I use `subpop_groups: "all"` - `config_sample_2pop_modifiers_test_random_subpop_groups.yml`- (which should force all subpopulations to have the same parameter value, at least for inference), then I get
- `snpi` file, `seir_modifier:modifiers:Ro_relax` - varying by slot but same across locations. 
- `hnpi` file, `outcome_modifiers:modifiers:test_limits` - varying but same across locations. 

which is what we expect. 

To summarize, yes this PR seems to fix the main underlying problem, but it has brought a behavior that needs to be documented better and discussed - during forward-only simulations, when `hpni`, `snpi`, and `hpni` are set to be drawn randomly from distributions, they very by slot and location, while `spar` varies only by slot. 

Thoughts from @jcblemai, @shauntruelove, @saraloo on the desired behavior here would be helpful.

pearsonca commented:
Nice investigation.

Agree we need to clearly document what matches / doesn't when stochastically simulating. Where in the gitbook do you think it's best to capture that?

Regarding behavior, all the stochastic features should match in the same way in my opinion. Weird that spar seems to deviate, so we should figure out why that is. I kind of think that location shouldn't cause a different sample either, but it's most important that we have consistent behavior.

jcblemai commented:
I agree Alison on consistency. The current state is not motivated by anything but historical hastly made decisions during COVID (fixed R0 across subpops, but age structure different in each state). We should change that carefully.

Now:
* @alsnhll  I don't see this config in the repository (perhaps I am in the wrong branch), does the outcome section has a parameter modifying file (another weird thing).
* @TimothyWillard An opinion: While this address the issue, the underlying problem is the seed issue (see my comment here https://github.com/HopkinsIDD/flepiMoP/issues/406#issuecomment-2501226752), and probably just resetting the distribution leaves other possible bugs. One would expect seeds to be different for each slot -- we can talk about handling that deterministically for reproducibility (but right now it is not done deterministically even if it is the same) but at least different simulations should not share seed. So on addition to allowing other random generators to still be in sync (e.g if one does stochastic simulations), it is pretty fuzzy (for me though, perhaps you got it) how it would work without being at the mercy of a scipy distribution implementation change (right now created distributions uses their own seed ? what happens here that makes the number different) ?
Even if we keep this update distribution approach, we need to seed each correctly with a different seed (at least like the emcee, ideally deterministically).

alsnhll commented:
Configs are :

alsnhll commented:
Configs are here:
https://github.com/HopkinsIDD/flepiMoP/blob/feature/406/random-draws-per-slot-for-SEIR/examples/tutorials/config_sample_2pop_modifiers_test_random.yml
https://github.com/HopkinsIDD/flepiMoP/blob/feature/406/random-draws-per-slot-for-SEIR/examples/tutorials/config_sample_2pop_modifiers_test_random_subpop_groups.yml

No parameter modifying outcome file

saraloo commented:
Hmm yes thanks for this investigation @alsnhll , agree with @jcblemai it's just historical decision making responsible for these inconsistencies. I'm not entirely sure what we should define here...

I see a few arguments either way
- though inconsistent, I think defining `spar` parameters to be the same across locations makes sense given our typical application of subpops and strata. If these are considered more as biological parameters (eg infectious period) it makes sense to me to keep them as consistent across locations. 
- Alternatively one could argue it's possible to think of these as not necessarily biological factors but indeed varying across location (eg contact parameters varying across populations). 
- following the same logic then, it's not immediately clear to me why the `hpar` parameters vary by location. How we define these are indeed a bit muddied - can be combination of both biological severity and reporting etc. 

Could use some discussion wrt how to harmonise these definitions. In any case, needs to be made clear in docs etc. 

> I don't know how to make hpar parameters be grouped by location - can we do this?

Similar inconsistencies in the npi parameters too as alison points out - ie we can group `seir_modifiers` but **not** `outcome_modifiers`. Again this was just development that was implemented at the time in response to something we needed to model, but can work on including this feature to all types of modifiers.

jcblemai commented on flepimop/gempyor_pkg/src/gempyor/outcomes.py:
```diff
@@ -91,6 +91,8 @@ def onerun_delayframe_outcomes(
     load_ID: bool = False,
     sim_id2load: int = None,
 ):
+    np.random.seed(seed=sim_id2write)
```
This might fix a seed in a way that is not always desirable: with classical inference, every ith iteration will have the same seed across slots. I would rely on /dev/random for now to be sure things are working correctly and keep the reproducibility of runs as a follow-up issue.

pearsonca replied:
that suggests a fix is needed in how classical inference works - any inference method should be managing set seeding in a sensible way.

jcblemai replied:
I fully agree. I would suggest a plan as follows:
1. Make thing works correctly as people expect them (so doing the real random thing that the emcee is doing also here)
2. Streamline the pathway to call gempyor (https://github.com/HopkinsIDD/flepiMoP/discussions/463#discussioncomment-11821663 shows 3 ways gempyor is called, but we should have one)
3. Build something with seed so that the run with the above pathway is reproducible.

If we do 3. before 2. there might be places that have unexpected behavior which is IMO more critical than being reproducible (even though that's also very important)

TimothyWillard replied:
Addressed in 48c1bf4c12486a8024e4804b5ba4ff0b4b7e65ab and c31aa840703fa932c7a69e026f04a8706833c277.

TimothyWillard commented:
I apologize for the confusion @pearsonca and @jcblemai, but this PR has not been ready for code review but now is. I was working through the GH action failure on c828f1e2a45e1d8395497669c823aa45ad1828ec, which was challenging to debug due to differences in behavior between Mac+Windows vs Linux. 

The quick version is both `gempyor.seir.run_parallel_SEIR` and `gempyor.outcomes.run_parallel_outcomes` are thin wrappers around their non-parallel counterparts that use [`tqdm.contrib.concurrent.process_map`](https://github.com/tqdm/tqdm/blob/0ed5d7f18fa3153834cbac0aa57e8092b217cc16/tqdm/contrib/concurrent.py#L72) to parallelize them. This function from `tqdm` creates a [`concurrent.futures.ProcessPoolExecutor`](https://docs.python.org/3.11/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor) with the [default multiprocessing start method](https://docs.python.org/3.11/library/multiprocessing.html#contexts-and-start-methods) which is `spawn` for Mac+Windows and `fork` for Linux. 

Reinitializing the parameters distributions was sufficient for addressing the original issue under `spawn`, but did not address the issue under `fork`. In fact, the unit tests suggest that sampling for `hpar` has not been working correctly on linux (like the HPCs). I've added the following unit tests in this PR to demonstrate these issues and allow for testing cross platform:

- Direct testing of `Parameters.reinitialize_distributions` which reset the random seed held onto by the random samplers in the `Parameters` class under spawn,
- Overall testing of the `flepimop simulate` CLI by running it and then manually checking the outputted files for the desired behavior, and
- (new since c828f1e2a45e1d8395497669c823aa45ad1828ec) testing of `gempyor.seir.run_parallel_SEIR` and `gempyor.outcomes.run_parallel_outcomes` by running a separate process, force setting the multiprocessing start method with [`multiprocessing.set_start_method`](https://docs.python.org/3.11/library/multiprocessing.html#multiprocessing.set_start_method), and then checking the outputted files (same as above point).

There is some duplicated code across these tests that I know I need to consolidate/clean up, but beyond that this is now ready for review. I will also note that I think this PR is a temporary fix to this problem, the correct longer term solution would be to [explicitly manage the generator that `numpy` uses to create the samples as `numpy` recommends](https://numpy.org/doc/stable/reference/random/generator.html) ([`np.random.seed` is considered legacy](https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html)).

@alsnhll @saraloo To clarify:

> So now variation in spar seems to be working as intended, and nothing has changed about how hpar, snpi, and hnpi work.

I'm taking this (and other commentary) to mean that we need similar tests for `snpi` and `hnpi` (I suspect that we may not be sampling correctly on Linux due to the above)? I think the usage of `subpop_groups: "all"` address the location variation issues, but to ensure this we probably need a unit test. Are there other values that this configuration key can take besides "all" that y'all know of? Does this configuration key apply to `hpar` and `spar`? Is this a correct summary of where we want things:

|        | Varies By Slot | Varies By Location By Default | Effect Of `subpop_groups: 'all'` |
|--------|----------------|-------------------------------|----------------------------------|
| `spar` | Yes            | No                            | N/A                              |
| `hpar` | Yes            | Yes                           | No Variation By Location         |
| `snpi` | Yes            | Yes                           | No Variation By Location         |
| `hnpi` | Yes            | Yes                           | No Variation By Location         |

Please feel free to correct me if I'm misunderstanding or if I've not got this table setup correctly (i.e. should be determined by some other variable or something).

jcblemai commented:
Impressive, and agree for the Generator thing.

In the meantime though, I would advise random seed from OS random source or something else instead of fixing seed on sim_id which has blind spots as sim_id is used differently in inference and local run. not that important though.

> I I think the usage of subpop_groups: "all" address the location variation issues, but to ensure this we probably need a unit test. Are there other values that this configuration key can take besides "all" that y'all know of? Does this configuration key apply to hpar and spar? Is this a correct summary of where we want things:

This key can be a list of list of subpop, which are then grouped together. e.g [["California", "Utah"], ["North Carolina", South Carolina"]] means that "California", "Utah" will have one value, and "North Carolina", South Carolina" another, and the rest of the subpopulation will have independent values. This is a key for modifiers

TimothyWillard commented:
This PR is now ready for review again. The biggest changes since the last comments are:

1. Unit tests for `snpi` and `hnpi` model outputs, and
2. Using explicit random seeds from the parallel caller rather than seeding based on the sim id.

The behavior is now:

|                             | Varies By Slot | Varies By Location |
|-----------------------------|----------------|--------------------|
| `spar`                      | Yes            | No                 |
| `hpar`                      | Yes            | Yes                |
| `snpi`                      | Yes            | Yes                |
| `hnpi`                      | Yes            | Yes                |
| `snpi` with `subpop_groups` | Yes            | No                 |
| `hnpi` with `subpop_groups` | Yes            | No                 |

A question that was discussed, but doesn't seem like we came to a strong conclusion on is "Should `hpar` vary by location by default?". If we want to change that I can work those changes into this PR as well.

jcblemai reviewed (approved):
This is nice, appreciate the seed handling. Very nice that we get same behavior. Does hpar varies per location even without an helper file ? In any case, I think hpar and spar should be harmonized and I'm agnostic to the choice taken.

jcblemai commented:
> > Does hpar varies per location even without an helper file?
> 
> I'm not sure what you mean by this, could you elaborate? What helper file are you referring to?

outcome config option `param_from_file` allow one to give a file that modifies each outcome parameters by multplying it with something (a somewhat old and inelegant interface, you could e.g use a county age-pyramid in there, that should be superseeded by something. I believe re-written modifiers are a good single-point to handle time and subpop variations of thing, as it handle all use cases -- though code is ugly.).

Also, to answer my question above: even without param_from_file, outcomes probabilities are "one draw per subpop" making them inconsistent with how spar are used.

saraloo commented:
I think harmonising `hpar` behaviour with `spar` makes the most sense to me, given how the config is defined etc?

jcblemai reviewed (approved):
All good, this is big change so should be part of a release.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/outcomes.py:
```diff
@@ -18,40 +19,33 @@
     sim_id2writes = np.arange(sim_id2write, sim_id2write + modinf.nslots)
-
-    loaded_values = None
+    random_seeds = np.random.choice(
+        range(nslots + 1, 4 * nslots), size=nslots, replace=False
+    ).tolist()
```
confirming:
 - this is ensuring we have different seeds across different slots
 - these seeds are themselves a random draw, with no way to ensure a particular draw
 - so we are guaranteeing different outcome draws for different invocations?

pearsonca replied:
can leave adding seed option for future issue generally addressing matching more broadly - might be via generators anyway, so :shrug:

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/outcomes.py:
```diff
@@ -124,6 +118,34 @@ def onerun_delayframe_outcomes(
         )
 
 
+def _onerun_delayframe_outcomes_with_random_seed(
+    random_seed: int,
+    sim_id2write: int,
+    modinf: model_info.ModelInfo,
+    load_ID: bool = False,
+    sim_id2load: int = None,
+) -> None:
+    """
+    Wrapper function to run `onerun_delayframe_outcomes` with a random seed.
+
+    Args:
+        random_seed: Random seed to use for the run.
+        sim_id2write: Simulation ID to write.
+        modinf: ModelInfo object.
+
+    Returns:
+        None
+
+    See Also:
+        `onerun_delayframe_outcomes`
+
+    """
+    np.random.seed(seed=random_seed)
+    onerun_delayframe_outcomes(
+        sim_id2write, modinf, load_ID=load_ID, sim_id2load=sim_id2load
+    )
```
This solution seems fine for now.

However, I foresee that we may have this need more generically: that is, to have a core function which can run with or without a random number set. This seems like an ideal use case for decorating with some sort of intercepting wrapper that will set the random seed if provided, otherwise proceed as normal

TimothyWillard replied:
I do not foresee this. Global seed setting with `numpy.random.seed` is considered legacy, see: https://numpy.org/doc/2.2/reference/random/generated/numpy.random.seed.html. Instead we should move to providing an explicit `numpy.random.Generator` in the future, see: https://github.com/HopkinsIDD/flepiMoP/pull/429#issuecomment-2590013488.

pearsonca replied:
Hmm. If not passing seed, then passing generator vs using "global" one. Should probably be that we're writing things in the other direction then - the explicit generator-passed-version is the real method, and the wrapping method is a version that will fill in the global generator (and probably issue some sort of warning)?

Relative to implementating distributions - my instinct is that we want *our* distribution classes to have our draw methods receive a generator instance. This is a bit more overarching random number generation overhaul, however.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/outcomes.py:
```diff
@@ -423,17 +445,16 @@ def compute_all_multioutcomes(
-                ].as_random_distribution()(
-                    size=len(modinf.subpop_struct.subpop_names)
-                )  # one draw per subpop
+                probabilities = np.repeat(
+                    parameters[new_comp]["probability"].as_random_distribution()(),
+                    len(modinf.subpop_struct.subpop_names),
+                )
```
can we get back the comment explaining what's happening here?

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/utils.py:
```diff
@@ -637,6 +637,29 @@ def list_filenames(
+    dfs: list[pd.DataFrame] | pd.DataFrame = []
+    for i, f in enumerate(sorted(files)):
+        dfs.append(pd.read_parquet(f) if f.suffix == ".parquet" else pd.read_csv(f))
+        dfs[-1]["slot"] = int(f.name.split(".")[0])
```
feels like we should have something like "extract_slot" function - that must be being done a few times in the code.

TimothyWillard replied:
Made note here https://github.com/HopkinsIDD/flepiMoP/issues/257#issuecomment-2767066186.

pearsonca commented on flepimop/gempyor_pkg/tests/outcomes/run_parallel_outcomes_test_script.py:
jumps out as a weird script. could use a bit of an explainer where its going to be used by other scripts (presumably for some multiple instance stuff?).

TimothyWillard replied:
The reason for this script existing is documented in the test that uses it. I'll also add documentation here.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/testing.py:
```diff
@@ -275,6 +281,62 @@ def sample_fits_distribution(
         return bool(np.greater(sample, 0.0))
 
 
+def run_test_in_separate_process(
+    script: str | Path, dest: Path | None = None, args: Iterable[str] = ()
+) -> None:
```
I don't think we need to return the whole subprocess, but seems reasonable to provide back the return code, *especially* if we some day want to actually use return codes and then want to test obtaining specific errors.

pearsonca commented on flepimop/gempyor_pkg/tests/outcomes/test_run_parallel_outcomes.py:
```diff
@@ -0,0 +1,74 @@
+    1. Sets up the test environment by copying the necessary files to a temporary directory.
+    2. Runs a pared down version of `gempyor.simulate.simulate` in a new process.
+    3. Reads the contents of the 'hpar' directory as a DataFrame.
+    4. Tests the contents of the 'hpar' DataFrame.
```
this could use more explanation as to what is being tested / why - this is about ensuring that each subpop gets ...unique values? the same values?

pearsonca commented on flepimop/gempyor_pkg/tests/outcomes/test_run_parallel_outcomes.py:
```diff
@@ -0,0 +1,74 @@
+            & (hpar["quantity"] == "probability")
+            & (hpar["outcome"] == "incidCase")
+        ]["value"].nunique()
+        == 10
```
code smell - where does this 10 come from? presumably from some element of the config - is there some way to get it from there for this comparison?

pearsonca commented on flepimop/gempyor_pkg/tests/outcomes/test_run_parallel_outcomes.py:
```diff
@@ -0,0 +1,74 @@
+            & (hpar["quantity"] == "probability")
+            & (hpar["outcome"] == "incidCase")
+        ]["value"].nunique()
+        == 10
```
ibid

pearsonca commented on flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py:
```diff
@@ -646,3 +665,36 @@ def test_parameters_reduce(self) -> None:
+    def test_reinitialize_parameters(self, tmp_path: pathlib.Path) -> None:
+        mock_inputs = distribution_three_valid_parameter_factory(tmp_path)
+
+        np.random.seed(123)
```
is this necessary? my read of the below is kinda that it should independent of the seed?

pearsonca commented on flepimop/gempyor_pkg/tests/parameters/test_parameters_class.py:
```diff
@@ -646,3 +665,36 @@ def test_parameters_reduce(self) -> None:
+
+        np.random.seed(123)
+
+        params = mock_inputs.create_parameters_instance()
+
+        with ProcessPoolExecutor(max_workers=2, mp_context=mp.get_context("spawn")) as ex:
+            results = list(
+                ex.map(
+                    sample_params,
+                    repeat(params, times=6),
+                    repeat(False, times=6),
+                )
+            )
+
+        for i in range(1, len(results)):
+            assert np.allclose(results[i - 1], results[i])
+
+        np.random.seed(123)
+
+        with ProcessPoolExecutor(max_workers=2, mp_context=mp.get_context("spawn")) as ex:
+            results_with_reinit = list(
+                ex.map(
+                    sample_params,
+                    repeat(params, times=6),
+                    repeat(True, times=6),
+                )
+            )
+
+        for i in range(1, len(results_with_reinit)):
+            assert not np.allclose(results_with_reinit[i - 1], results_with_reinit[i])
```
if the seed doesn't actually matter, feels like these can all be folded together?

pearsonca commented on flepimop/gempyor_pkg/tests/seir/run_parallel_SEIR_test_script.py:
ibid re comment on outcomes.

also feels like there should be a way to combine the outcomes and this? isn't outcomes just this + some extra work?

pearsonca reviewed (commented):
Several questions. Maybe some stuff to fix.

TimothyWillard commented on flepimop/gempyor_pkg/src/gempyor/outcomes.py:
```diff
@@ -18,40 +19,33 @@
 def run_parallel_outcomes(
     modinf: model_info.ModelInfo, *, sim_id2write, nslots=1, n_jobs=1
-):
+) -> Literal[1]:
```
https://github.com/HopkinsIDD/flepiMoP/issues/536

TimothyWillard commented:
Todo list for this PR extracted from @pearsonca's comments:

- [ ] Add `extract_slot` function, refactor `read_directory` to use that function, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021441515.
- [ ] Document `tests/outcomes/run_parallel_outcomes_test_script.py` auxiliary script, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021448117.
- [ ] Change the return of `gempyor.testing.run_test_in_separate_process` to the return code of the executed script, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021451474 and https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021452603.
- [ ] Include an explanation of "what is being tests / why" in `tests/outcomes/test_run_parallel_outcomes.py`, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021457311.
- [ ] Remove hard-coded value in unit tests, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021458403, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021458661, and https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021472486.
- [ ] Remove(?) `np.random.seed` call in the `test_reinitialize_parameters` test case in `tests/parameters/test_parameters_class` and then refactor duplicate code afterwards, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021463096 and https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021464094.
- [ ] Consolidate `tests/outcomes/run_parallel_outcomes_test_script.py` and `tests/seir/run_parallel_SEIR_test_script.py` auxiliary scripts into one unified script, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021467874.
- [ ] Add comments to `tests/simulate/test_random_seir_parameter_draw_per_slot.py` to tests, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021511313.
- [ ] Change `4 * nslots`/`4 * modinf.nslots` magic in `run_parallel_outcomes`/`run_parallel_SEIR` to be larger to ensure more randomness, https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021602574.
- [ ] Document/unit test that successive calls to `run_parallel_outcomes`/`run_parallel_SEIR` will produce different outcomes (i.e. draw different random seeds), https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021602574.
- [ ] Rework unit tests to not use separate configurations for `config_sample_2pop_modifiers_test_random.yml` and `config_sample_2pop_modifiers_test_random_subpop_groups.yml` (I guess add call back to modify config in memory before writing to temporary location in `gempyor.testing.setup_example_from_tutorials`?), https://github.com/HopkinsIDD/flepiMoP/pull/429#discussion_r2021399710.

TimothyWillard commented on examples/tutorials/config_sample_2pop_modifiers_test_random.yml:
Created targeted issue for this one, GH-544

pearsonca reviewed (approved):
LGTM

TimothyWillard commented:
The behavior is now:

|                             | Varies By Slot | Varies By Location |
|-----------------------------|----------------|--------------------|
| `spar`                      | Yes            | No                 |
| `hpar`                      | Yes            | No                 |
| `snpi`                      | Yes            | Yes                |
| `hnpi`                      | Yes            | Yes                |
| `snpi` with `subpop_groups` | Yes            | No                 |
| `hnpi` with `subpop_groups` | Yes            | No                 |
